### PR TITLE
store_client: new module to handle authenticated requests (CRAFT-445)

### PR DESCRIPTION
**Pull request by sergiusens**
Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

-------

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: craft_store/store_client.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        :param permissions: Set of permissions to grant the login.
+        :param description: Client description to refer to from the Store.
+        :param ttl: time to live for the credential, in other words, how
+                    long until it expires, expressed in seconds.
+        """
+        token_request = self._endpoints.get_token_request(
+            permissions=permissions, description=description, ttl=ttl
+        )
+
+        token = self._get_token(token_request)
```
Comment (cmatsuoka) [inline craft_store/store_client.py:118]: Maybe call this variable `macaroon` to make its nature more explicit?

**Review by cmatsuoka — APPROVED** [review]
I'm not familiar with candid operations, but the overall request authentication looks good to me.

File: craft_store/store_client.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        )
+
+        return token_exchange_response.json()["macaroon"]
+
+    def login(
+        self,
+        *,
+        permissions: Sequence[str],
+        description: str,
+        ttl: str,
```
Comment (matiasb) [inline craft_store/store_client.py:99]: FYI, an integer value is expected in the API.

**Review by matiasb — APPROVED** [review]
LGTM

File: craft_store/store_client.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    def _candid_discharge(self, macaroon: str) -> str:
+        bakery_macaroon = bakery.Macaroon.from_dict(json.loads(macaroon))
+        discharges = bakery.discharge_all(
+            bakery_macaroon, self._bakery_client.acquire_discharge
+        )
+
+        # serialize macaroons the bakery-way
+        discharged_macaroons = (
+            "[" + ",".join(map(utils.macaroon_to_json_string, discharges)) + "]"
```
Comment (facundobatista) [inline craft_store/store_client.py:78]: Consider re-implementing `macaroon_to_json_string` so we don't need to use a private module from the lib. The function is actually one line, and it uses "public" helpers.

File: craft_store/store_client.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        )
+
+        return base64.urlsafe_b64encode(utils.to_bytes(discharged_macaroons)).decode(
```
Comment (facundobatista) [inline craft_store/store_client.py:81]: This "to_bytes" is actually a `.encode("utf8")`, as we're sure it's unicode.

**Review by facundobatista — CHANGES_REQUESTED** [review]
Annotated a couple of details.

Also, I do not understand in which step the user *actually logs in*...

File: craft_store/store_client.py (hunk 1 above)
Comment (sergiusens) [reply craft_store/store_client.py]: updated in https://github.com/canonical/craft-store/pull/6/commits/84fa660f874ab10b86dcf59eb1fa68c141c64234

File: craft_store/store_client.py (hunk 4 above)
Comment (sergiusens) [reply craft_store/store_client.py]: updated in https://github.com/canonical/craft-store/pull/6/commits/e12ac2076348199eff3867e11575cdbdf107d67c

File: craft_store/store_client.py (hunk 3 above)
Comment (sergiusens) [reply craft_store/store_client.py]: updated in https://github.com/canonical/craft-store/pull/6/commits/e12ac2076348199eff3867e11575cdbdf107d67c

File: craft_store/store_client.py (hunk 2 above)
Comment (sergiusens) [reply craft_store/store_client.py]: addressed in https://github.com/canonical/craft-store/pull/6/commits/9f0a81a1446734197e0379d539be477ea7e544d2

**Review by facundobatista — APPROVED** [review]
Approving, but I still do not understand in which step the user actually logs in...

**Comment by sergiusens** [comment]
@facundobatista  I think I understood your question now, the webview to login is brought up when running `bakery.discharge_all`

**Pull request merged**